### DISCUSSION TITLE
Float rice compression is no longer an option

### DISCUSSION THREAD

**Discussion by at88mph** [nom-tam-fits/nom-tam-fits#389] [Q&A]
I'm using an existing Rice compression file (https://github.com/nom-tam-fits/nom-tam-fits/blob/master/src/test/resources/nom/tam/image/provided/m13_rice.fits) to test streaming cutouts from compressed images.  It contains a `ZBITPIX` that translates to a `float` primitive (`float.class`) when looking up the `ICompressorControl`:

```java
final String quantizeAlgorithmName = null;
final String compressionAlgorithmName = "RICE_1";
final Class<?> primitiveClass = float.class;
final ICompressorControl compressorControl = 
    CompressorProvider.findCompressorControl(quantizeAlgorithmName, compressionAlgorithmName, primitiveClass);
// compressorControl is null
```

I noticed that the `RiceCompressor.FloatRiceCompressor` is no more (grant it, it was just a wrapper around the `IntRiceCompressor`), but that means there is no Float support.  Is there another way to do this?

Thanks!

**Comment by at88mph** [comment]
As a follow up, after looking through the code more, if I set the `quantizeAlgorithmName` to `float.class.getName()`, it seems to work.  Should there be a conditional for how that parameter is set?  I normally pull the `ZQUANTIZ` header for that.

**Comment by at88mph** [comment]
Sorry to spam this conversation, but I found one more thing.  According to the FITS standard paper here (https://fits.gsfc.nasa.gov/standard40/fits_standard40draft1.pdf), if the `ZQUANTIZ` is absent, then the default should be `NO_DITHER`.  If I default to that:

```java
getHeader().getStringValue(Compression.ZQUANTIZ, Compression.ZQUANTIZ_NO_DITHER);
```

My tests pass.  Is this the appropriate way to do this?

Thank you, again.

**Reply by attipaci** [reply]
That certainly looks OK to me. However, you should never really need to handle `ZQUANTIZ` or any other compression header keywords manually. These should be handled automatically by the compression classes, e.g.  by the `.asImageHDU()` method of the `CompressedImageHDU` class... As far as I recall these already do default to `NO_DITHER` if the `ZQUANTIZ` keyword is absent. If you want to use this for e.g. handling decompression of individual tiles, you can see how `ZQUANTIZ` is handled for entire images, and try to mimic that if possible. Did that answer the question?

**Reply by at88mph** [reply]
Thanks @attipaci .  The problem I'm trying to fix is that the `asImageHDU()` decompresses the entire image into memory, which won't work for our 75GB cubes.  Instead, our cutout library will make use of decompressing only those tiles that overlap and streaming out the desired pixels.  This will be my next MR.

**Comment by attipaci** [comment]
Hi Dustin! I'll look into it. The compression support of the library is my biggest nightmare. I'm having a hard time figuring out what Ritchie had in mind when adding it. I'll see what I can make of your questions. Give me a few days... (and feel free to remind me if I don't get back to you in a while...)

**Comment by attipaci** [comment]
As for the first question about `FloatRiceCompressor`. The real reason this class was removed is because it worked with a special option class that had the baked in quantization option. However, quantization and compression algorithms are two separate things (or steps), which were treated together on a case-by-case basis unfortunately, rather than in a general manner.... That said, I'll see if I can restore back compatibility by spoofing the old classes and method signatures with the reorganized compression hierarchy... (I added a new issue #390 for it -- and will try to resolve it for 1.18.0...)

The (conceptually) better way to do float Rice compression is to handle the options for quantization (float to int conversion) separately from those of the compression itself (int only), e.g. as:

```java
  var riceOpt = new RiceCompressOption();      // options for RICE algorithm
  var quantOpt = new QuantizeOption(riceOpt);  // quantization options (with pass-through options for the compression algorithm)
  
  // Now set quantization and algorithmic options as desired
  ...
  
  // And then use these options to create a quant+rice compressor....
  var floatRice = new QuantizeProcessor.FloatQuantCompressor(quantOpt, new RiceCompressor.IntRiceCompressor(riceOpt));
  floatRice.compress(...);
```

It's not as elegant as it should be, with the pass-through algorithm options for the quantized compressor class (confusingly named `QuantizeProcessor` instead of say QuantizeCompressor), but this is as far as it goes without re-structuring compression from the ground up.

**Comment by attipaci** [comment]
I have reinstated `FloatRiceCompressor` and the `DoubleRiceCompressor` classes and corresponding options in PR #392, now merged to master. The same goes for `FloatHCompressor` and `DoubleHCompressor`.

**Comment by at88mph** [comment]
Alright, thank you @attipaci.  I meant to say that defaulting to the `NO_DITHER` option seems to be working alright, but with this change I can remove that.  Thanks.

**Reply by attipaci** [reply]
It's good to have options, lol. Your question made me realize that I did not have a compelling reason for removing those classes/methods just because the compression stuff underwent open-heart surgery. So I am just grateful that you brought me to that realization ;-).
